psa-crypto/src/operations/key_management.rs

```diff
@@ -227,17 +219,87 @@ pub fn export_public(key: Id, data: &mut [u8]) -> Result<usize> {
     let handle = key.handle()?;
     let mut data_length = 0;
 
-    Status::from(unsafe {
+    let export_res = Status::from(unsafe {
         psa_crypto_sys::psa_export_public_key(
             handle,
             data.as_mut_ptr(),
             data.len(),
             &mut data_length,
         )
     })
-    .to_result()?;
+    .to_result();
+    key.close_handle(handle)?;
+    export_res?;
+    Ok(data_length)
+}
 
+/// Gets the attributes for a given key ID
+///
+/// The `Id` structure can be created with the `from_persistent_key_id` constructor on `Id`.
+///
+/// # Example
+///
+/// ```
+/// # use psa_crypto::operations::key_management;
+/// # use psa_crypto::types::key::{Attributes, Type, Lifetime, Policy, UsageFlags};
+/// # use psa_crypto::types::algorithm::{AsymmetricSignature, Hash};
+/// # let mut attributes = Attributes {
+/// #     key_type: Type::RsaKeyPair,
+/// #     bits: 1024,
+/// #     lifetime: Lifetime::Volatile,
+/// #     policy: Policy {
+/// #         usage_flags: UsageFlags {
+/// #             sign_hash: true,
+/// #             sign_message: true,
+/// #             verify_hash: true,
+/// #             verify_message: true,
+/// #             ..Default::default()
+/// #         },
+/// #         permitted_algorithms: AsymmetricSignature::RsaPkcs1v15Sign {
+/// #             hash_alg: Hash::Sha256.into(),
+/// #         }.into(),
+/// #     },
+/// # };
+/// psa_crypto::init().unwrap();
+/// let my_key = key_management::generate(attributes, None).unwrap();
+/// //...
+/// let key_attributes = key_management::get_key_attributes(my_key);
+/// ```
+
+pub fn get_key_attributes(key: Id) -> Result<Attributes> {
+    initialized()?;
+    let mut key_attributes = unsafe { psa_crypto_sys::psa_key_attributes_init() };
+    let handle = key.handle()?;
+    let attributes_res = Status::from(unsafe {
+        psa_crypto_sys::psa_get_key_attributes(handle, &mut key_attributes)
+    })
+    .to_result();
     key.close_handle(handle)?;
+    attributes_res?;
+    Ok(Attributes::try_from(key_attributes)?)
+}
 
-    Ok(data_length)
+/// Completes a new key operation (either generate or import)
+///
+/// If key is not `Volatile` (`Persistent` or `Custom(u32)`), handle is closed.
+///
+/// If a key is `Volatile`, `Id` returned contains the key `handle`. Otherwise, it does not.
+fn complete_new_key_operation(
+    key_lifetime: Lifetime,
+    id: psa_key_id_t,
+    handle: psa_key_handle_t,
+    operation_result: Result<()>,
+) -> Result<Id> {
+    if key_lifetime != Lifetime::Volatile {
+        Status::from(unsafe { psa_crypto_sys::psa_close_key(handle) }).to_result()?;
+    }
+    operation_result?;
```

Comment:
Do you think that if `psa_generate_key` or `psa_import_key`  failed you still need to close the key? I would have thought that if those operations fail, they do not open the key.
The documentation of `psa_generate_key` says:
```
 * \param[out] handle       On success, a handle to the newly created key.
 *                          \c 0 on failure.
```
So I think this is fine not passing `operation_result` to this function (checking in the caller if it failed or not).

Reply:
I have checked `psa_generate_key` and `psa_import_key` - you are correct, a handle should not be created unless the operation succeeds.